include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -546,7 +546,31 @@ make_transform_iterator(_Iter __it, _UnaryFunc __unary_func)
     return transform_iterator<_Iter, _UnaryFunc>(__it, __unary_func);
 }
 
-template <typename SourceIterator, typename IndexMap>
+namespace __internal
+{
+
+//functor concept
+struct __functor_concept
+{
+    template <typename _T>
+    static auto
+    test(int) -> decltype(::std::declval<_T>()(0), ::std::true_type{});
+
+    template <typename _T>
+    static auto
+    test(...) -> ::std::false_type;
+};
+
+template <typename _T>
+inline constexpr bool __is_functor = decltype(__functor_concept::test<_T>(0))::value;
+
+} // namespace __internal
+
+template <typename SourceIterator, typename _Permutation,
+          typename IndexMap = typename std::conditional<
+              !__internal::__is_functor<_Permutation>, _Permutation,
+              transform_iterator<counting_iterator<typename ::std::iterator_traits<SourceIterator>::difference_type>,
+                                 _Permutation>>::type>
```

Comment:
I do not like the extra template argument in the public API. Does it really need to be there (as opposed to e.g. a typedef within the class)?

Reply:
Yes, I'll move it within the class.